File: sort.js
// Start of the program
const fs = require('fs/promises')
const { performance } = require('perf_hooks')

async function loadArrayData() {
	try {
		const data = await fs.readFile('./inputs/randomNumbersArray.json', { encoding: 'utf8' })
		return JSON.parse(data)
	} catch (err) {
		console.error('Reading file error:', err)
	}
}

// BUBBLE SORT
function bubbleSort(array) {
	//Outer pass
	for (let i = 0; i < array.length; i++) {
		//Inner pass
		for (let j = 0; j < array.length - i - 1; j++) {
			//Comparing values in ascending order
			if (array[j + 1] < array[j]) {
				//Swapping values
				[array[j + 1], array[j]] = [array[j], array[j + 1]]
			}
		}
	}
	return array
}

// MERGE SORT
function merge(left, right) {
	const array = []
	// Break out of loop if any one of the array gets empty
	while (left.length && right.length) {
		// Pick the smaller among the smallest element of left and right sub arrays 
		if (left[0] < right[0]) {
			array.push(left.shift())
		} else {
			array.push(right.shift())
		}
	}
	// Concatenating the leftover elements
	// (in case we didn't go through the entire left or right array)
	return [...array, ...left, ...right]
}

function mergeSort(array) {
	const half = array.length / 2
	// Base case for exiting recursion
	if (array.length < 2) {
		return array
	}
	const left = array.splice(0, half)
	return merge(mergeSort(left), mergeSort(array))
}

function saveResult(method, array) {
	const filePath = `./outputs/node${method}.json`
	fs.writeFile(filePath, JSON.stringify(array, null, 4), err => {
		if (err) console.error('Writing file error: ', err)
	}).then(() => console.log(`Sorted array by method = ${method} saved here ${filePath} !`))
}

function main() {
	loadArrayData().then(data => {
		const array = data
		const array2 = [...array]
		// BUBBLE SORT
		console.log('Bubble Sort Started!')
		console.time('BubbleSort')
		const bubbleResult = bubbleSort(array)
		console.timeEnd('BubbleSort')
		// MERGE SORT
		console.log('Merge Sort Started!')
		console.time('MergeSort')
		const mergeResult = mergeSort(array2)
		console.timeEnd('MergeSort')
		// Save sorted arrays
		saveResult('BubbleSort', bubbleResult)
		saveResult('MergeSort', mergeResult)
	})
}

main()

File: sort.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"time"
)

// BUBBLE SORT
func BubbleSort(array []int) []int {
	for i := 0; i < len(array); i++ {
		for j := 0; j < len(array)-i-1; j++ {
			if array[j] > array[j+1] {
				array[j], array[j+1] = array[j+1], array[j]
			}
		}
	}
	return array
}

// MERGE SORT
func merge(left []int, right []int) []int {
	final := []int{}
	i := 0
	j := 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			final = append(final, left[i])
			i++
		} else {
			final = append(final, right[j])
			j++
		}
	}
	for ; i < len(left); i++ {
		final = append(final, left[i])
	}
	for ; j < len(right); j++ {
		final = append(final, right[j])
	}
	return final
}

func mergeSort(array []int) []int {
	if len(array) < 2 {
		return array
	}
	left := mergeSort(array[:len(array)/2])
	right := mergeSort(array[len(array)/2:])
	return merge(left, right)
}

func saveResults(method string, array []int) {
	var filePath string = "./outputs/go" + method + ".json"
	file, _ := json.MarshalIndent(array, "", " ") //this parse the array into a writable object
	err := ioutil.WriteFile(filePath, file, 0644)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Sorted array by method = %s saved here %s !", method, filePath)
	fmt.Println()
}

func main() {
	data, err := ioutil.ReadFile("./inputs/randomNumbersArray.json")
	if err != nil {
		log.Fatal(err)
	}

	var numbers []int
	var numbers2 []int
	json.Unmarshal(data, &numbers)
	json.Unmarshal(data, &numbers2)
	// BUBBLE SORT
	bubbleStart := time.Now()
	bubbleResult := BubbleSort(numbers)
	fmt.Printf("Bubble sort %s", time.Since(bubbleStart))
	fmt.Println()
	// MERGE SORT
	mergeStart := time.Now()
	mergeResult := mergeSort(numbers2)
	fmt.Printf("Merge Sort %s", time.Since(mergeStart))
	fmt.Println()
	// Save sort results
	saveResults("BubbleSort", bubbleResult)
	saveResults("MergeSort", mergeResult)
}
